src/main/java/org/prebid/server/vast/VastModifier.java

```diff
@@ -102,43 +110,55 @@ private static String resolveVastXmlFrom(String bidAdm, String bidNurl) {
         }
         throw new PreBidException("VastXml does not contain neither InLine nor Wrapper for %s response"
                 .formatted(bidder));
     }
 
-    private static String appendTrackingUrl(String vastXml, String vastUrlTracking, String elementCloseTag) {
-        if (vastXml.contains(IMPRESSION_CLOSE_TAG)) {
-            return insertAfterExistingImpressionTag(vastXml, vastUrlTracking);
+    private static String appendTrackingUrl(String vastXml, String vastUrlTracking, Matcher closeTagMatcher) {
+        final Matcher impressionMatcher = IMPRESSION_CLOSE_TAG_PATTERN.matcher(vastXml);
+        if (impressionMatcher.find()) {
+            return insertAfterExistingImpressionTag(vastXml, vastUrlTracking, impressionMatcher);
         }
-        return insertBeforeElementCloseTag(vastXml, vastUrlTracking, elementCloseTag);
+        return insertBeforeElementCloseTag(vastXml, vastUrlTracking, closeTagMatcher);
     }
 
-    private static String insertAfterExistingImpressionTag(String vastXml, String vastUrlTracking) {
+    private static String insertAfterExistingImpressionTag(String vastXml,
+                                                           String vastUrlTracking,
+                                                           Matcher impressionMatcher) {
+
         final String impressionTag = "<Impression><![CDATA[" + vastUrlTracking + "]]></Impression>";
-        final int replacementStart = vastXml.lastIndexOf(IMPRESSION_CLOSE_TAG);
 
-        return vastXml.substring(0, replacementStart) + IMPRESSION_CLOSE_TAG + impressionTag
-                + vastXml.substring(replacementStart + IMPRESSION_CLOSE_TAG.length());
+        String group = impressionMatcher.group();
+        while (impressionMatcher.find()) {
+            group = impressionMatcher.group();
+        }
+        final int replacementStart = vastXml.lastIndexOf(group);
+
+        return vastXml.substring(0, replacementStart + group.length()) + impressionTag
+                + vastXml.substring(replacementStart + group.length());
     }
 
-    private static String insertBeforeElementCloseTag(String vastXml, String vastUrlTracking, String elementCloseTag) {
-        final int indexOfCloseTag = StringUtils.indexOfIgnoreCase(vastXml, elementCloseTag);
+    private static String insertBeforeElementCloseTag(String vastXml,
+                                                      String vastUrlTracking,
+                                                      Matcher closeTagMatcher) {
 
-        if (indexOfCloseTag == -1) {
+        if (!closeTagMatcher.find()) {
             return vastXml;
         }
 
+        final String group = closeTagMatcher.group();
+        final int indexOfCloseTag = vastXml.indexOf(group);
+
         final String caseSpecificCloseTag =
-                vastXml.substring(indexOfCloseTag, indexOfCloseTag + elementCloseTag.length());
+                vastXml.substring(indexOfCloseTag, indexOfCloseTag + group.length());
```

Comment:
Let's get rid of unnecessary searches and replacements:
- `vastXml.indexOf(group)` -> `matcher.start()`
- `vastXml.lastIndexOf(group)` -> `matcher.end()`
- `vastXml.replace(tag, impressionTag + tag)` -> `insertAt(...)`

General idea:
```
    private static String appendTrackingUrlToVastXml(String xml, String urlTracking, String bidder) {
        return appendTrackingUrl(xml, urlTracking, INLINE_OPEN_TAG_PATTERN, INLINE_CLOSE_TAG_PATTERN)
                .or(() -> appendTrackingUrl(xml, urlTracking, WRAPPER_OPEN_TAG_PATTERN, WRAPPER_CLOSE_TAG_PATTERN))
                .orElseThrow(() -> new PreBidException(
                        "VastXml does not contain neither InLine nor Wrapper for %s response".formatted(bidder)));
    }

    private static Optional<String> appendTrackingUrl(String vastXml,
                                                      String vastUrlTracking,
                                                      Pattern openTagPattern,
                                                      Pattern closeTagPattern) {

        final Matcher openTagMatcher = openTagPattern.matcher(vastXml);
        if (!openTagMatcher.find()) {
            return Optional.empty();
        }

        final Matcher impressionMatcher = IMPRESSION_CLOSE_TAG_PATTERN.matcher(vastXml);
        if (impressionMatcher.find()) {
            return Optional.of(insertAfterExistingImpressionTag(vastXml, vastUrlTracking, impressionMatcher));
        }

        final int openTagEnd = openTagMatcher.end();
        final Matcher closeTagMatcher = closeTagPattern.matcher(vastXml);
        if (!closeTagMatcher.find(openTagEnd)) {
            return Optional.of(vastXml);
        }

        return Optional.of(insertUrlTracking(vastXml, closeTagMatcher.start(), vastUrlTracking));
    }

    private static String insertAfterExistingImpressionTag(String vastXml,
                                                           String vastUrlTracking,
                                                           Matcher impressionMatcher) {

        int replacementEnd = impressionMatcher.end();
        while (impressionMatcher.find()) {
            replacementEnd = impressionMatcher.end();
        }

        return insertUrlTracking(vastXml, replacementEnd, vastUrlTracking);
    }

    private static String insertUrlTracking(String vastXml, int index, String vastUrlTracking) {
        final String impressionTag = "<Impression><![CDATA[" + vastUrlTracking + "]]></Impression>";
        return vastXml.substring(0, index) + impressionTag + vastXml.substring(index);
    }
```